pkg/image/api/validation/validation.go

```diff
@@ -59,9 +62,104 @@ func validateImage(image *api.Image, fldPath *field.Path) field.ErrorList {
 		}
 	}
 
+	for i, sig := range image.Signatures {
+		result = append(result, validateImageSignature(&sig, fldPath.Child("signatures").Index(i))...)
+	}
+
 	return result
 }
 
+func validateImageSignature(signature *api.ImageSignature, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(signature.Type) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("type"), ""))
+	}
+	if len(signature.Content) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("content"), ""))
+	}
+	if len(signature.ForImage) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("forImage"), ""))
+	}
+	if len(signature.State) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("state"), ""))
+	}
+
+	allowedValues := sets.NewString(
+		string(api.SignatureForImageUnknown),
+		string(api.SignatureForImageMatch),
+		string(api.SignatureForImageMismatch))
+	if !allowedValues.Has(string(signature.ForImage)) {
+		msg := fmt.Sprintf("invalid value of ForImage, acceptable values are: %s", strings.Join(allowedValues.List(), ", "))
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("forImage"), signature.ForImage, msg))
+	}
+
+	allowedValues = sets.NewString(
+		string(api.SignatureStateUnknown),
+		string(api.SignatureStateTrusted),
+		string(api.SignatureStateRevoked),
+		string(api.SignatureStateExpired),
+		string(api.SignatureStateUntrusted))
+	if !allowedValues.Has(string(signature.State)) {
+		msg := fmt.Sprintf("invalid value for State, acceptable values are: %s", strings.Join(allowedValues.List(), ", "))
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("state"), signature.State, msg))
+	}
+
+	switch signature.ForImage {
+	case api.SignatureForImageUnknown:
+		if signature.State != api.SignatureStateUnknown && signature.State != api.SignatureStateUntrusted {
+			msg := fmt.Sprintf("invalid value of ForImage for known signature state %q", signature.State)
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("forImage"), signature.ForImage, msg))
+		}
+	case api.SignatureForImageMismatch:
+		if signature.State != api.SignatureStateUntrusted {
+			msg := fmt.Sprintf("invalid value of ForImage for signature state %q", signature.State)
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("forImage"), signature.ForImage, msg))
+		}
```

Comment:
This will eventually (not right now) have to be relaxed or expanded with extra states, to support trusted signatures matching a base image but not the top-level image (“this application was built on top of a RHEL image and here is a signature to prove it”).